### PR TITLE
If 373 asset details maintenance

### PR DESCRIPTION
**Pull request by g1glh0rn**

### CODE REVIEW

**Comment by mattmikulina** [comment]
Task linked: [IF-373 Maintenance of an asset (Asset view)](https://app.clickup.com/t/2177764/IF-373)

File: fusionfrontend/src/app/components/content/maintenance-progressbar/maintenance-progressbar.component.scss [hunk 1]
```diff
@@ -55,7 +54,7 @@
 }
```
Comment (jsylupp) [inline fusionfrontend/src/app/components/content/maintenance-progressbar/maintenance-progressbar.component.scss:54]: add color-variables everywhere

File: fusionfrontend/src/app/dashboards/components/content/maintenance-list/maintenance-list.component.ts [hunk 2]
```diff
@@ -316,10 +249,25 @@ export class MaintenanceListComponent implements OnInit, OnChanges {
     return openAlertPriority;
   }
 
+  isLastChildElement(rowNode: any): boolean {
+    const subsystemIds = rowNode.parent?.data.subsystemIds;
+    if (subsystemIds) {
+      const index = subsystemIds.findIndex((value) => value === rowNode.node.data.id);
+      return index === subsystemIds.length - 1;
+    } else {
+      return null;
```
Comment (jsylupp) [inline fusionfrontend/src/app/dashboards/components/content/maintenance-list/maintenance-list.component.ts:258]: return false?

File: fusionfrontend/src/app/dashboards/components/content/maintenance-list/maintenance-list.component.ts [hunk 3]
```diff
@@ -333,13 +281,55 @@ export class MaintenanceListComponent implements OnInit, OnChanges {
...
-      return MaintenanceState.MEDIUMTERM;
+  private isMaintenanceNeededSoonForMaintenanceType(asset: FactoryAssetDetailsWithFields, type: MaintenanceType) {
+    return Utils.getMaintenanceValue(asset, type)
+      && Utils.getMaintenancePercentage(asset, type) < this.MAINTENANCE_HIGHLIGHT_PERCENTAGE;
+  }
+
+  private updateTree() {
+    if (this.displayedFactoryAssets) {
+      const expandedNodeIDs = this.getExpandedNodeIDs(this.treeData);
+      const map = this.displayedFactoryAssets.map(asset => asset.subsystemIds);
```
Comment (jsylupp) [inline fusionfrontend/src/app/dashboards/components/content/maintenance-list/maintenance-list.component.ts:292]: are there more specific names for map and reduce?

File: fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.html [hunk 4]
```diff
@@ -15,119 +15,42 @@
 
 <div class='asset-details-wrapper'>
```
Comment (jsylupp) [inline fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.html:16]: correct class name for this component?

File: fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.scss [hunk 5]
```diff
@@ -16,74 +16,100 @@
...
-    .pointslabel{
-      margin: 0px;
+    .pointslabel {
+      margin: 0;
     }
   }
 
-  .dater{
-    margin: 0px;
+  .dater {
```
Comment (jsylupp) [inline fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.scss:45]: typo?

File: fusionfrontend/src/app/factory/services/factory-resolver.service.ts [hunk 6]
```diff
@@ -15,7 +15,7 @@
 
 import { Injectable } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
-import { BehaviorSubject, forkJoin, Observable, Subject } from 'rxjs';
+import { BehaviorSubject, combineLatest, forkJoin, Observable, Subject } from 'rxjs';
 import { map, mergeMap, switchMap } from 'rxjs/operators';
 // import { switchMap } from 'rxjs/operators';
```
Comment (jsylupp) [inline fusionfrontend/src/app/factory/services/factory-resolver.service.ts:20]: remove

File: fusionfrontend/src/app/store/asset/asset.service.ts [hunk 7]
```diff
@@ -186,4 +192,28 @@ export class AssetService {
     mappedAsset.connectionString = assetDetails.connectionString;
     return mappedAsset;
   }
+
+  updateAssetWithFieldValue(asset: FactoryAssetDetailsWithFields, secondsInPast: number) {
```
Comment (jsylupp) [inline fusionfrontend/src/app/store/asset/asset.service.ts:196]: add return type.  Maybe move to oispService?

File: fusionfrontend/src/app/factory/services/factory-resolver.service.ts [hunk 8]
```diff
@@ -116,6 +118,16 @@ export class FactoryResolver {
             assetDetailsArray.map(assetDetails => this.fieldService.getFieldsOfAsset(companyId, assetDetails.id))))
       ).subscribe();
       this.assetsWithDetailsAndFields$ = this.factoryComposedQuery.joinAssetsDetailsWithFieldInstancesWithAlerts();
+      this.assetsWithDetailsAndFieldsAndValues$ = this.assetsWithDetailsAndFields$.pipe(
+        mergeMap((assets) =>
+          combineLatest(
+            assets.map((asset) => {
+              return this.assetService.updateAssetWithFieldValue(asset, 600);
```
Comment (jsylupp) [inline fusionfrontend/src/app/factory/services/factory-resolver.service.ts:125]: just a question: Why 600?  Might this be changed later on?

**Review by jsylupp — CHANGES_REQUESTED** [review]

File: fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.scss [hunk 9]
```diff
@@ -16,74 +16,100 @@
...
+.asset-details-wrapper {
+  border: 0;
   background: $light-grey;
-  margin: 0px;
+  margin: 0;
   padding: 20px;
 
-  .modal-body{
+  .modal-body {
     height: 170px;
```
Comment (jsylupp) [inline fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.scss:26]: consider using rem

File: fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.html [hunk 10]
```diff
@@ -15,119 +15,42 @@
...
-      </div>
-      <div class="radio btn">
-          <input type="radio" name="btn-group-demo-radios" id="btn-demo-radio-4" value="customDate" [(ngModel)]="timeSlotOptions" (click)="setOptions('customDate', false)">
-          <label for="btn-demo-radio-4">Custom Date</label>
-      </div>
+  <ng-container *ngIf="asset$ | async as assetWithFields">
+
+    <div class="caption">Maintenance</div>
+    <div class="progress-bar-container">
+      <div class="progress-bar-label">Operating Hours till Maintenance</div>
```
Comment (jsylupp) [inline fusionfrontend/src/app/factory/components/pages/asset-details/asset-performance/asset-performance.component.html:22]: could be changed to <label>

**Pull request merged**